tests/src/Test/Tasty/Vivado/GenTcl.hs

```diff
@@ -0,0 +1,154 @@
+
+instance Semigroup SimProj where
+  (<>) (SimProj hdl0 ip0) (SimProj hdl1 ip1) =
+    SimProj (hdl0<>hdl1) (ip0<>ip1)
+
+instance Monoid SimProj where
+  mempty = SimProj [] []
+
+thread :: [a -> a] -> a -> a
+thread = foldr (.) id
+
+-- | Given something like @topEntity_dcfifo_240A2A9D3FF8D64B.tcl@, return the
+-- IP name, @dcfifo@
+parseTcl :: FilePath -> String
+parseTcl fp = takeWhile (/= '_') (drop 1 $ dropWhile (/= '_') fp)
+
+mSimProj :: FilePath -> T.Text -> FilePath -> SimProj -> SimProj
+mSimProj dir lib fp p@(SimProj hdl ip)
+  | ".vhdl" `isSuffixOf` fp || ".v" `isSuffixOf` fp = SimProj ((lib, dir </> fp):hdl) ip
+  | ".tcl" `isSuffixOf` fp = SimProj hdl ((dir </> fp):ip)
+  | otherwise = p
+
+-- | From @PortProductsSum_testBench@, extract @testBench@ (for instance)
+stripEntity ::
+  T.Text ->
+  T.Text
+stripEntity entityName =
+  let parts = T.split (=='_') entityName
+    in T.intercalate "_" (filter (isLower . T.head) parts)
+
+simProjFromClashEntities ::
+  -- | HDL source dir
+  FilePath ->
+  -- | Qualified name
+  String ->
+  IO SimProj
+simProjFromClashEntities hdlDir qualName = do
+  [(_,mHdl)] <- getManifests (hdlDir </> qualName </> "clash-manifest.json")
+  let deps = T.unpack <$> transitiveDependencies mHdl
+  nextSimProj <- traverse (simProjFromClashEntities hdlDir) deps
+  pure $ mconcat nextSimProj <> asSimProj qualName mHdl
+
+-- | Read @clash-manifest.json@ and create a TCL script to instantiate relevant
+-- IP, read sources.
+tclFromManifest ::
+  HdlSource ->
+  -- | Module name
+  String ->
+  -- | Entity
+  String ->
+  -- | TCL script
+  IO String
+tclFromManifest src moduleName entity = do
+  proj <- simProjFromClashEntities (directory src) (moduleName ++ "." ++ hsEntity)
+  cabalDir <- getDataDir
+  pure $ mkTcl cabalDir entity src proj
+ where
+  hsEntity = T.unpack (stripEntity (T.pack entity))
+
+asSimProj :: FilePath -> Manifest -> SimProj
+asSimProj dir m =
+  let fps = fst <$> fileNames m
+    in thread (mSimProj dir (topComponent m) <$> fps) mempty
+
+mkTcl ::
+  FilePath ->
+  -- | Cabal @data@ dir
+  String ->
+  -- | Entity
+  HdlSource ->
+  SimProj ->
+  String
+mkTcl cabalDir entity hdl (SimProj hdls tcls) =
+  [i|
+set_msg_config -severity {CRITICAL WARNING} -new_severity ERROR
+
+file mkdir ip
+file mkdir project
+
+source #{cabalDir}/data-files/tcl/clash_namespace.tcl
+
+namespace eval myspace {}
+
+set myspace::hdl_dir #{hdlDir}
+
+|]
+
+  ++ concatMap (\tcl ->
+      [i|clash::loadTclIface #{parseTcl tcl} ${myspace::hdl_dir}/#{tcl}\n|]) tcls
+
+  ++
+      [i|
+create_project -in_memory
+set myspace::ips [clash::runClashScripts]
+if {$myspace::ips ne {}} {
+  set myspace::ipFiles [get_property IP_FILE [get_ips $myspace::ips]]
+}
+close_project
+
+if {$myspace::ips ne {}} {
+  read_ip $myspace::ipFiles
+  generate_target {synthesis simulation} [get_ips $myspace::ips]
+}
+|]
+
+  ++ readSources
+
+  ++ [i|
+update_compile_order -fileset [current_fileset]
```

Comment:
Could you add a comment here stating that we need to do this due to a bug?

Reply:
Yes!